internal/node/machine_mgr.go

```diff
@@ -347,6 +345,11 @@ func (m *MachineManager) awaitHandshake(vmid string) {
 	for !handshakeOk && !m.stopping() {
 		if time.Now().UTC().After(timeoutAt) {
 			m.log.Error("Did not receive NATS handshake from agent within timeout.", slog.String("vmid", vmid))
+			if len(m.handshakes) == 0 {
+				m.log.Error("First handshake failed, shutting down to avoid inconsistent behavior")
+				m.cancel()
+				os.Exit(1)
```

Comment:
You can also drop this call to `os.Exit()` - the cancel function will result in a 0 exit code but the error will be logged.